python/lib/core/dmod/core/context/base.py

```diff
@@ -0,0 +1,197 @@
+    """
+    Checks to see if a member of an object is a property
+
+    Args:
+        obj: The object to check
+        member_name: The member on the object to check
+
+    Returns:
+        True if the member with the given name on the given object is a property
+    """
+    if not hasattr(obj, member_name):
+        raise AttributeError(f"{obj} has no attribute '{member_name}'")
+
+    if isinstance(obj, type):
+        return isinstance(getattr(obj, member_name), property)
+
+    # Is descriptor: inspect.isdatadescriptor(dict(inspect.getmembers(obj.__class__))[member_name])
+    parent_reference = dict(inspect.getmembers(obj.__class__))[member_name]
+    return isinstance(parent_reference, property)
+
+
+@typing.runtime_checkable
+class ObjectCreatorProtocol(typing.Protocol):
+    """
+    Defines the bare minimum methods that will be used that may create objects
+    """
+    @abc.abstractmethod
+    def create_object(self, name: str, /, *args, **kwargs) -> T:
+        """
+        Create an object and store its reference
+
+        Args:
+            name: The name of the class to instantiate
+            *args: Positional arguments used to instantiate the object
+            **kwargs: Keyword arguments used to instantiate the object
+
+        Returns:
+            A proxy pointing at the instantiated object
+        """
+
+    @abc.abstractmethod
+    def __str__(self):
+        ...
+
+    @abc.abstractmethod
+    def __repr__(self):
+        ...
+
+
+class ObjectManagerScope(abc.ABC, ObjectCreatorProtocol):
+    """
+    Maintains references to objects that have been instantiated via an object manager within a specific scope
+    """
+    def __init__(self, name: str, logger: LoggerProtocol = None):
+        self.__name = name
+        self.__items: typing.List[typing.Any] = []
+        self.__scope_id: uuid.UUID = uuid.uuid1()
+        self.__on_close: typing.List[
+            typing.Union[
+                typing.Callable[[ObjectManagerScope], typing.Any],
+                typing.Callable[[], typing.Any]
+            ]
+        ] = []
+        # Record the 4th frame as to where this scope started
+        # 0 = The `format_stack_trace` function
+        # 1 = This constructor
+        # 2 = The scope subclass' constructor
+        # 3 = Where the object was instantiated
+        self.__started_at: str = format_stack_trace(3)
+        self.__logger: LoggerProtocol = logger or logging.getLogger()
+
+    def _perform_on_close(self, handler: typing.Union[typing.Callable[[ObjectManagerScope], typing.Any], typing.Callable[[], typing.Any]]):
+        if not isinstance(handler, typing.Callable):
+            raise ValueError(
+                f"The handler passed to {self.__class__.__name__}._perform_on_close must be a some sort of function "
+                f"but got {type(handler)} instead"
+            )
+
+    @abc.abstractmethod
+    def create_object(self, name: str, /, *args, **kwargs) -> T:
```

Comment:
It returns `T` rather than `Any` due to how some tooling handle dynamic type hinting. The goal here is to instantiate an object that is a proxy but pretend it's not one. I forgot the exact conditions to trigger where it really helps, but it has a downstream meta-benefit that doesn't help the code, it just helps IDEs with referencing objects further down the line.  `typing.Any` also often tells tooling to stop trying to figure stuff out.

Reply:
I think you might have to really color outside the lines to have an object reference the scope as its the scope on the client said that references the object, not the other way around.

Do you want me to look at adding in some guard rails?